File: sdk-node/test/sign.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { createHmac } from 'node:crypto';
import { signRequest, newNonce, HDR_APPLICATION_ID, HDR_TIMESTAMP, HDR_NONCE, HDR_SIGNATURE } from '../src/sign.js';

// Pinned known vector — must match sdk-go/sign_test.go's TestSignRequest_KnownVector.
// If this hex drifts, an SDK is producing a canonical form the server
// won't accept.
test('signRequest produces the known-vector hex', () => {
  const method = 'POST';
  const path = '/v1/users';
  const body = '{"email":"alice@example.com"}';
  const appId = 'app_test';
  const secret = 'topsecret';
  const nonce = 'abc123';
  const ts = 1704067200;

  const headers = signRequest(method, path, body, appId, secret, ts * 1000, nonce);

  assert.equal(headers[HDR_APPLICATION_ID], appId);
  assert.equal(headers[HDR_TIMESTAMP], String(ts));
  assert.equal(headers[HDR_NONCE], nonce);
  assert.equal(
    headers[HDR_SIGNATURE],
    '92c2aca00ba47b4377aebf3e3af134aff93cf7ad959d05e31a0f0618df9f7d9a',
  );

  // Belt and braces: recompute the way the server would.
  const canonical = `${method}|${path}|${ts}|${nonce}|${body}`;
  const want = createHmac('sha256', secret).update(canonical).digest('hex');
  assert.equal(headers[HDR_SIGNATURE], want);
});

test('signRequest handles empty body', () => {
  const h = signRequest('GET', '/v1/users', '', 'a', 's', 1000_000, 'n1');
  assert.equal(h[HDR_SIGNATURE].length, 64);
});

test('newNonce returns unique 32-char hex strings', () => {
  const seen = new Set<string>();
  for (let i = 0; i < 100; i++) {
    const n = newNonce();
    assert.equal(n.length, 32);
    assert.match(n, /^[0-9a-f]{32}$/);
    assert.equal(seen.has(n), false);
    seen.add(n);
  }
});

File: sdk-go/sign_test.go
package ztxbas

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"net/http"
	"strings"
	"testing"
	"time"
)

// TestSignRequest_KnownVector verifies that the SDK produces the exact
// canonical form and hex digest the server's HMAC middleware expects.
// If this test ever drifts, RPs will start seeing INVALID_SIGNATURE —
// so pinning it here catches accidental changes early.
func TestSignRequest_KnownVector(t *testing.T) {
	const (
		appID  = "app_test"
		secret = "topsecret"
		nonce  = "abc123"
		body   = `{"email":"alice@example.com"}`
		method = "POST"
		path   = "/v1/users"
	)
	// Pinned timestamp: 2024-01-01T00:00:00Z = 1704067200
	now := time.Unix(1704067200, 0)
	ts := "1704067200"

	req, err := http.NewRequest(method, "https://example.com"+path, strings.NewReader(body))
	if err != nil {
		t.Fatalf("build request: %v", err)
	}
	signRequest(req, []byte(body), appID, secret, now, nonce)

	// Recompute the expected signature the same way the server does.
	canonical := fmt.Sprintf("%s|%s|%s|%s|%s", method, path, ts, nonce, body)
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(canonical))
	want := hex.EncodeToString(mac.Sum(nil))

	if got := req.Header.Get(hdrSignature); got != want {
		t.Errorf("signature mismatch:\n got  %s\n want %s", got, want)
	}
	if got := req.Header.Get(hdrApplicationID); got != appID {
		t.Errorf("app id header: got %q want %q", got, appID)
	}
	if got := req.Header.Get(hdrTimestamp); got != ts {
		t.Errorf("timestamp header: got %q want %q", got, ts)
	}
	if got := req.Header.Get(hdrNonce); got != nonce {
		t.Errorf("nonce header: got %q want %q", got, nonce)
	}
	// Signature is deterministic HMAC — pin the hex so a canonical
	// form change fails loudly here rather than silently at runtime.
	const wantHex = "92c2aca00ba47b4377aebf3e3af134aff93cf7ad959d05e31a0f0618df9f7d9a"
	if want != wantHex {
		t.Errorf("expected known-vector hex mismatch\n got  %s\n want %s", want, wantHex)
	}
}

func TestSignRequest_EmptyBody(t *testing.T) {
	req, err := http.NewRequest("GET", "https://example.com/v1/users", nil)
	if err != nil {
		t.Fatalf("build request: %v", err)
	}
	signRequest(req, nil, "app_x", "sec", time.Unix(1000, 0), "n1")
	sig := req.Header.Get(hdrSignature)
	if len(sig) != 64 {
		t.Errorf("signature length = %d, want 64", len(sig))
	}
}

func TestNewNonce_UniqueAndHex(t *testing.T) {
	seen := make(map[string]bool)
	for i := 0; i < 100; i++ {
		n, err := newNonce()
		if err != nil {
			t.Fatalf("newNonce: %v", err)
		}
		if len(n) != nonceBytes*2 {
			t.Fatalf("nonce length = %d, want %d", len(n), nonceBytes*2)
		}
		if _, err := hex.DecodeString(n); err != nil {
			t.Errorf("nonce not hex: %v", err)
		}
		if seen[n] {
			t.Errorf("duplicate nonce %q", n)
		}
		seen[n] = true
	}
}
